Assertion violation: undefined `relativeSpecifiers` on `Object.getModuleSpecifiers`

From @dlannoye:

```
dependencies_path\typescript@4.2.1-rc\node_modules\typescript\lib\tsc.js:91576
                    ts.Debug.checkDefined(relativeSpecifiers);
                             ^
Error: Debug Failure.
    at Object.getModuleSpecifiers (dependencies_path\typescript@4.2.1-rc\node_modules\typescript\lib\tsc.js:91576:30)
    at getSpecifierForModuleSymbol (dependencies_path\typescript@4.2.1-rc\node_modules\typescript\lib\tsc.js:39998:62)
    at dependencies_path\typescript@4.2.1-rc\node_modules\typescript\lib\tsc.js:39874:39
    at Array.map (<anonymous>)
    at getSymbolChain (dependencies_path\typescript@4.2.1-rc\node_modules\typescript\lib\tsc.js:39872:58)
    at lookupSymbolChainWorker (dependencies_path\typescript@4.2.1-rc\node_modules\typescript\lib\tsc.js:39858:51)
    at lookupSymbolChain (dependencies_path\typescript@4.2.1-rc\node_modules\typescript\lib\tsc.js:39852:24)
    at symbolToTypeNode (dependencies_path\typescript@4.2.1-rc\node_modules\typescript\lib\tsc.js:40005:29)
    at typeToTypeNodeHelper (dependencies_path\typescript@4.2.1-rc\node_modules\typescript\lib\tsc.js:39150:28)
    at mapToTypeNodes (dependencies_path\typescript@4.2.1-rc\node_modules\typescript\lib\tsc.js:39671:40)

```

@flybayer I can’t reproduce with your repo. I get a few checker errors but no crash:

```
❯ yarn tsc
yarn run v1.22.4
$ /System/Volumes/Data/Developer/microsoft/eg/tsc-broken/node_modules/.bin/tsc
app/auth/mutations/forgotPassword.test.ts:12:3 - error TS2698: Spread types may only be created from object types.

12   ...jest.requireActual("blitz")!,
     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

db/index.ts:2:10 - error TS2305: Module '"@prisma/client"' has no exported member 'PrismaClient'.

2 import { PrismaClient } from "@prisma/client"
           ~~~~~~~~~~~~

types.ts:3:10 - error TS2614: Module '"db"' has no exported member 'User'. Did you mean to use 'import User from "db"' instead?

3 import { User } from "db"
           ~~~~


Found 3 errors.

error Command failed with exit code 2.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

@dlannoye (or actually anyone else who has seen this error), are you using pnpm or some other system that puts modules in a hidden folder (starting with `.`) inside `node_modules`? The line you reverted to fix your problem is supposed to filter out invalid module specifiers, and the crash means you were left with _no_ valid module specifiers after the filtering. But if you’re not using pnpm, I’m very confused as to what was getting filtered out in the first place. If anyone can provide a small reproduction of this bug, that would be very helpful.

--- a/src/compiler/moduleSpecifiers.ts
+++ b/src/compiler/moduleSpecifiers.ts
@@ -282,12 +282,12 @@ namespace ts.moduleSpecifiers {
         const redirects = host.redirectTargetsMap.get(importedPath) || emptyArray;
         const importedFileNames = [...(referenceRedirect ? [referenceRedirect] : emptyArray), importedFileName, ...redirects];
         const targets = importedFileNames.map(f => getNormalizedAbsolutePath(f, cwd));
-        let filterIgnoredPaths = some(targets, not(containsIgnoredPath));
+        let shouldFilterIgnoredPaths = !every(targets, containsIgnoredPath);
 
         if (!preferSymlinks) {
             // Symlinks inside ignored paths are already filtered out of the symlink cache,
             // so we only need to remove them from the realpath filenames.
-            const result = forEach(targets, p => !(filterIgnoredPaths && containsIgnoredPath(p)) && cb(p, referenceRedirect === p));
+            const result = forEach(targets, p => !(shouldFilterIgnoredPaths && containsIgnoredPath(p)) && cb(p, referenceRedirect === p));
             if (result) return result;
         }
         const links = host.getSymlinkCache
@@ -314,13 +314,13 @@ namespace ts.moduleSpecifiers {
                 for (const symlinkDirectory of symlinkDirectories) {
                     const option = resolvePath(symlinkDirectory, relative);
                     const result = cb(option, target === referenceRedirect);
-                    filterIgnoredPaths = true; // We found a non-ignored path in symlinks, so we can reject ignored-path realpaths
+                    shouldFilterIgnoredPaths = true; // We found a non-ignored path in symlinks, so we can reject ignored-path realpaths
                     if (result) return result;
                 }
             });
         });
         return result || (preferSymlinks
-            ? forEach(targets, p => filterIgnoredPaths && containsIgnoredPath(p) ? undefined : cb(p, p === referenceRedirect))
+            ? forEach(targets, p => shouldFilterIgnoredPaths && containsIgnoredPath(p) ? undefined : cb(p, p === referenceRedirect))
             : undefined);
     }
 

--- a/src/compiler/moduleSpecifiers.ts
+++ b/src/compiler/moduleSpecifiers.ts
@@ -282,10 +282,12 @@ namespace ts.moduleSpecifiers {
         const redirects = host.redirectTargetsMap.get(importedPath) || emptyArray;
         const importedFileNames = [...(referenceRedirect ? [referenceRedirect] : emptyArray), importedFileName, ...redirects];
         const targets = importedFileNames.map(f => getNormalizedAbsolutePath(f, cwd));
+        let filterIgnoredPaths = some(targets, not(containsIgnoredPath));
+
         if (!preferSymlinks) {
             // Symlinks inside ignored paths are already filtered out of the symlink cache,
             // so we only need to remove them from the realpath filenames.
-            const result = forEach(targets, p => !containsIgnoredPath(p) && cb(p, referenceRedirect === p));
+            const result = forEach(targets, p => !(filterIgnoredPaths && containsIgnoredPath(p)) && cb(p, referenceRedirect === p));
             if (result) return result;
         }
         const links = host.getSymlinkCache
@@ -312,12 +314,13 @@ namespace ts.moduleSpecifiers {
                 for (const symlinkDirectory of symlinkDirectories) {
                     const option = resolvePath(symlinkDirectory, relative);
                     const result = cb(option, target === referenceRedirect);
+                    filterIgnoredPaths = true; // We found a non-ignored path in symlinks, so we can reject ignored-path realpaths
                     if (result) return result;
                 }
             });
         });
         return result || (preferSymlinks
-            ? forEach(targets, p => containsIgnoredPath(p) ? undefined : cb(p, p === referenceRedirect))
+            ? forEach(targets, p => filterIgnoredPaths && containsIgnoredPath(p) ? undefined : cb(p, p === referenceRedirect))
             : undefined);
     }
 

--- a/tests/baselines/reference/nodeModuleReexportFromDottedPath.js
+++ b/tests/baselines/reference/nodeModuleReexportFromDottedPath.js
@@ -1,0 +1,33 @@
+//// [tests/cases/compiler/nodeModuleReexportFromDottedPath.ts] ////
+
+//// [index.d.ts]
+export interface PrismaClientOptions {
+  rejectOnNotFound?: any;
+}
+
+export class PrismaClient<T extends PrismaClientOptions = PrismaClientOptions> {
+  private fetcher;
+}
+
+//// [index.d.ts]
+export * from ".prisma/client";
+
+//// [index.ts]
+import { PrismaClient } from "@prisma/client";
+declare const enhancePrisma: <TPrismaClientCtor>(client: TPrismaClientCtor) => TPrismaClientCtor & { enhanced: unknown };
+const EnhancedPrisma = enhancePrisma(PrismaClient);
+export default new EnhancedPrisma();
+
+
+//// [index.js]
+"use strict";
+exports.__esModule = true;
+var client_1 = require("@prisma/client");
+var EnhancedPrisma = enhancePrisma(client_1.PrismaClient);
+exports["default"] = new EnhancedPrisma();
+
+
+//// [index.d.ts]
+import { PrismaClient } from "@prisma/client";
+declare const _default: PrismaClient<import(".prisma/client").PrismaClientOptions>;
+export default _default;

--- a/tests/baselines/reference/nodeModuleReexportFromDottedPath.symbols
+++ b/tests/baselines/reference/nodeModuleReexportFromDottedPath.symbols
@@ -1,0 +1,41 @@
+=== /node_modules/.prisma/client/index.d.ts ===
+export interface PrismaClientOptions {
+>PrismaClientOptions : Symbol(PrismaClientOptions, Decl(index.d.ts, 0, 0))
+
+  rejectOnNotFound?: any;
+>rejectOnNotFound : Symbol(PrismaClientOptions.rejectOnNotFound, Decl(index.d.ts, 0, 38))
+}
+
+export class PrismaClient<T extends PrismaClientOptions = PrismaClientOptions> {
+>PrismaClient : Symbol(PrismaClient, Decl(index.d.ts, 2, 1))
+>T : Symbol(T, Decl(index.d.ts, 4, 26))
+>PrismaClientOptions : Symbol(PrismaClientOptions, Decl(index.d.ts, 0, 0))
+>PrismaClientOptions : Symbol(PrismaClientOptions, Decl(index.d.ts, 0, 0))
+
+  private fetcher;
+>fetcher : Symbol(PrismaClient.fetcher, Decl(index.d.ts, 4, 80))
+}
+
+=== /node_modules/@prisma/client/index.d.ts ===
+export * from ".prisma/client";
+No type information for this code.
+No type information for this code.=== /index.ts ===
+import { PrismaClient } from "@prisma/client";
+>PrismaClient : Symbol(PrismaClient, Decl(index.ts, 0, 8))
+
+declare const enhancePrisma: <TPrismaClientCtor>(client: TPrismaClientCtor) => TPrismaClientCtor & { enhanced: unknown };
+>enhancePrisma : Symbol(enhancePrisma, Decl(index.ts, 1, 13))
+>TPrismaClientCtor : Symbol(TPrismaClientCtor, Decl(index.ts, 1, 30))
+>client : Symbol(client, Decl(index.ts, 1, 49))
+>TPrismaClientCtor : Symbol(TPrismaClientCtor, Decl(index.ts, 1, 30))
+>TPrismaClientCtor : Symbol(TPrismaClientCtor, Decl(index.ts, 1, 30))
+>enhanced : Symbol(enhanced, Decl(index.ts, 1, 100))
+
+const EnhancedPrisma = enhancePrisma(PrismaClient);
+>EnhancedPrisma : Symbol(EnhancedPrisma, Decl(index.ts, 2, 5))
+>enhancePrisma : Symbol(enhancePrisma, Decl(index.ts, 1, 13))
+>PrismaClient : Symbol(PrismaClient, Decl(index.ts, 0, 8))
+
+export default new EnhancedPrisma();
+>EnhancedPrisma : Symbol(EnhancedPrisma, Decl(index.ts, 2, 5))
+

--- a/tests/baselines/reference/nodeModuleReexportFromDottedPath.types
+++ b/tests/baselines/reference/nodeModuleReexportFromDottedPath.types
@@ -1,0 +1,35 @@
+=== /node_modules/.prisma/client/index.d.ts ===
+export interface PrismaClientOptions {
+  rejectOnNotFound?: any;
+>rejectOnNotFound : any
+}
+
+export class PrismaClient<T extends PrismaClientOptions = PrismaClientOptions> {
+>PrismaClient : PrismaClient<T>
+
+  private fetcher;
+>fetcher : any
+}
+
+=== /node_modules/@prisma/client/index.d.ts ===
+export * from ".prisma/client";
+No type information for this code.
+No type information for this code.=== /index.ts ===
+import { PrismaClient } from "@prisma/client";
+>PrismaClient : typeof PrismaClient
+
+declare const enhancePrisma: <TPrismaClientCtor>(client: TPrismaClientCtor) => TPrismaClientCtor & { enhanced: unknown };
+>enhancePrisma : <TPrismaClientCtor>(client: TPrismaClientCtor) => TPrismaClientCtor & { enhanced: unknown; }
+>client : TPrismaClientCtor
+>enhanced : unknown
+
+const EnhancedPrisma = enhancePrisma(PrismaClient);
+>EnhancedPrisma : typeof PrismaClient & { enhanced: unknown; }
+>enhancePrisma(PrismaClient) : typeof PrismaClient & { enhanced: unknown; }
+>enhancePrisma : <TPrismaClientCtor>(client: TPrismaClientCtor) => TPrismaClientCtor & { enhanced: unknown; }
+>PrismaClient : typeof PrismaClient
+
+export default new EnhancedPrisma();
+>new EnhancedPrisma() : PrismaClient<import("/node_modules/.prisma/client/index").PrismaClientOptions>
+>EnhancedPrisma : typeof PrismaClient & { enhanced: unknown; }
+

--- a/tests/cases/compiler/nodeModuleReexportFromDottedPath.ts
+++ b/tests/cases/compiler/nodeModuleReexportFromDottedPath.ts
@@ -1,0 +1,19 @@
+// @declaration: true
+
+// @Filename: /node_modules/.prisma/client/index.d.ts
+export interface PrismaClientOptions {
+  rejectOnNotFound?: any;
+}
+
+export class PrismaClient<T extends PrismaClientOptions = PrismaClientOptions> {
+  private fetcher;
+}
+
+// @Filename: /node_modules/@prisma/client/index.d.ts
+export * from ".prisma/client";
+
+// @Filename: /index.ts
+import { PrismaClient } from "@prisma/client";
+declare const enhancePrisma: <TPrismaClientCtor>(client: TPrismaClientCtor) => TPrismaClientCtor & { enhanced: unknown };
+const EnhancedPrisma = enhancePrisma(PrismaClient);
+export default new EnhancedPrisma();

